vernac/vernacentries.ml

```diff
@@ -849,17 +849,48 @@ let private_ind =
   | Some () -> return true
   | None -> return false
 
+(** Flag governing use of primitive projections. Disabled by default. *)
+let primitive_flag =
+  Goptions.declare_bool_option_and_ref
+    ~depr:false
+    ~key:["Primitive";"Projections"]
+    ~value:false
+
+let primitive_proj =
+  let open Attributes in
+  let open Notations in
+  qualify_attribute "projections" (bool_attribute ~name:"primitive")
+  >>= function
+  | Some t -> return t
+  | None -> return (primitive_flag ())
+
 let vernac_inductive ~atts kind indl =
   let udecl, indl = extract_inductive_udecl indl in
   let is_defclass = match kind, indl with
   | Class _, [ ( id , bl , c , Constructors [l]), [] ] -> Some (id, bl, c, l)
   | _ -> None
   in
-  let ((template, (poly, cumulative)), private_ind), typing_flags = Attributes.(
+  let finite = finite_of_kind kind in
+  let is_record = function
+  | ((_ , _ , _ , RecordDecl _), _) -> true
+  | _ -> false
+  in
+  let is_constructor = function
+  | ((_ , _ , _ , Constructors _), _) -> true
+  | _ -> false
+  in
+  (* We only allow the #[projections(primitive)] attribute
+     in special cases. *)
+  let prim_proj_attr : bool Attributes.Notations.t =
+    if List.for_all is_record indl then primitive_proj
+    else Notations.return false
```

Comment:
I'm not an expert of these parsing combinators, maybe @SkySkimmer can tell us if this is the right way to fail

Reply:
This works OK, it will say something like "this command does not support this attribute: projections".
If you want a more customized error you can do like polymorphic_cumulative, passing a bool to primitive_proj so that it errors in the `Some` case.